[BUG] Internal error if a `header` parameter has `type: integer`

**Describe the bug**

For example, we have a parameter:

```
{"name": "X-Id", "in": "header", "required": True, "type": "integer"}
```

Then running Schemathesis will cause the following internal error:

```
value = 0

    def is_latin_1_encodable(value: str) -> bool:
        """Header values are encoded to latin-1 before sending."""
        try:
>           value.encode("latin-1")
E           AttributeError: 'int' object has no attribute 'encode'

```

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -8,6 +8,10 @@ Changelog
 
 - Improve the error message when the application can not be loaded from the value passed to the ``--app`` command-line option. `#836`_
 
+**Fixed**
+
+- Internal error when a "header" / "cookie" parameter were not coerced to a string before filtration. `#839`_
+
 `2.7.1`_ - 2020-10-22
 ---------------------
 
@@ -1436,6 +1440,7 @@ Deprecated
 .. _0.3.0: https://github.com/schemathesis/schemathesis/compare/v0.2.0...v0.3.0
 .. _0.2.0: https://github.com/schemathesis/schemathesis/compare/v0.1.0...v0.2.0
 
+.. _#839: https://github.com/schemathesis/schemathesis/issues/839
 .. _#836: https://github.com/schemathesis/schemathesis/issues/836
 .. _#824: https://github.com/schemathesis/schemathesis/issues/824
 .. _#816: https://github.com/schemathesis/schemathesis/issues/816

--- a/src/schemathesis/specs/openapi/serialization.py
+++ b/src/schemathesis/specs/openapi/serialization.py
@@ -107,6 +107,8 @@ def _serialize_query_openapi3(
 def _serialize_header_openapi3(
     name: str, type_: str, explode: Optional[bool]
 ) -> Generator[Optional[Callable], None, None]:
+    # Headers should be coerced to a string so we can check it for validity later
+    yield to_string(name)
     # Header parameters always use the "simple" style, that is, comma-separated values
     if type_ == "array":
         yield delimited(name, delimiter=",")
@@ -120,6 +122,8 @@ def _serialize_header_openapi3(
 def _serialize_cookie_openapi3(
     name: str, type_: str, explode: Optional[bool]
 ) -> Generator[Optional[Callable], None, None]:
+    # Cookies should be coerced to a string so we can check it for validity later
+    yield to_string(name)
     # Cookie parameters always use the "form" style
     if explode and type_ in ("array", "object"):
         # `explode=true` doesn't make sense
@@ -140,15 +144,19 @@ def _serialize_swagger2(definitions: DefinitionList) -> Generator[Optional[Calla
         name = definition["name"]
         collection_format = definition.get("collectionFormat", "csv")
         type_ = definition.get("type")
-        if definition["in"] != "body" and type_ in ("array", "object"):
-            if collection_format == "csv":
-                yield delimited(name, delimiter=",")
-            if collection_format == "ssv":
-                yield delimited(name, delimiter=" ")
-            if collection_format == "tsv":
-                yield delimited(name, delimiter="\t")
-            if collection_format == "pipes":
-                yield delimited(name, delimiter="|")
+        if definition["in"] == "header":
+            # Headers should be coerced to a string so we can check it for validity later
+            yield to_string(name)
+        if definition["in"] != "body":
+            if type_ in ("array", "object"):
+                if collection_format == "csv":
+                    yield delimited(name, delimiter=",")
+                if collection_format == "ssv":
+                    yield delimited(name, delimiter=" ")
+                if collection_format == "tsv":
+                    yield delimited(name, delimiter="\t")
+                if collection_format == "pipes":
+                    yield delimited(name, delimiter="|")
 
 
 serialize_openapi3_parameters = make_serializer(_serialize_openapi3)
@@ -309,3 +317,9 @@ def matrix_object(item: Generated, name: str, explode: Optional[bool]) -> None:
 def nothing(item: Generated, name: str) -> None:
     """Remove a key from an item."""
     item.pop(name, None)
+
+
+@conversion
+def to_string(item: Generated, name: str) -> None:
+    """Convert the value to a string."""
+    item[name] = str(item[name])

--- a/test/test_parameters.py
+++ b/test/test_parameters.py
@@ -10,34 +10,42 @@ import schemathesis
 from .utils import as_param
 
 
-def test_headers(testdir):
+@pytest.mark.parametrize("schema_name", ("simple_swagger.yaml", "simple_openapi.yaml"))
+@pytest.mark.parametrize("type_", ("string", "integer", "array", "boolean", "number"))
+def test_headers(testdir, schema_name, type_):
     # When parameter is specified for "header"
+    if schema_name == "simple_swagger.yaml":
+        kwargs = {"type": type_}
+    else:
+        kwargs = {"schema": {"type": type_}}
     testdir.make_test(
         """
 @schema.parametrize()
-@settings(suppress_health_check=[HealthCheck.filter_too_much], deadline=None)
+@settings(suppress_health_check=[HealthCheck.filter_too_much], deadline=None, max_examples=3)
 def test_(case):
-    assert_str(case.headers["api_key"])
+    assert_str(case.headers["X-Header"])
     assert_requests_call(case)
         """,
-        **as_param({"name": "api_key", "in": "header", "required": True, "type": "string"}),
+        schema_name=schema_name,
+        **as_param({"name": "X-Header", "in": "header", "required": True, **kwargs}),
     )
     # Then the generated test case should contain it in its `headers` attribute
     testdir.run_and_assert(passed=1)
 
 
-def test_cookies(testdir):
+@pytest.mark.parametrize("type_", ("string", "integer", "array", "object", "boolean", "number"))
+def test_cookies(testdir, type_):
     # When parameter is specified for "cookie"
     testdir.make_test(
         """
 @schema.parametrize()
-@settings(suppress_health_check=[HealthCheck.filter_too_much], deadline=None)
+@settings(suppress_health_check=[HealthCheck.filter_too_much], deadline=None, max_examples=3)
 def test_(case):
     assert_str(case.cookies["token"])
     assert_requests_call(case)
         """,
         schema_name="simple_openapi.yaml",
-        **as_param({"name": "token", "in": "cookie", "required": True, "schema": {"type": "string"}}),
+        **as_param({"name": "token", "in": "cookie", "required": True, "schema": {"type": type_}}),
     )
     # Then the generated test case should contain it in its `cookies` attribute
     testdir.run_and_assert(passed=1)

